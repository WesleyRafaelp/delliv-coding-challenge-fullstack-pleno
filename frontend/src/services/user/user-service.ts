import { HttpClient } from "../../infra/http-client/http-client";
import { tokenService } from "../auth/token-service";

export const userService = {
  async getProfile(user: any, ctx = null) {
    const token = tokenService.get(ctx);
    return HttpClient(
      `${process.env.NEXT_PUBLIC_URL_API}/users/${user.data.sub}`,
      {
        method: "GET",
      headers: { "Authorization": `Bearer ${token}` },
      }
    ).then(async (res) => {
      const body = await res.body;
      if (res.ok === false) {
        return { errors: body.message };
      }
      return { data: body };
    });
  },
  async create(user: any) {
    return await fetch(`${process.env.NEXT_PUBLIC_URL_API}/users`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(user),
    }).then(async (res) => {
      const body = await res.json();
      if (res.ok === false) {
        return { errors: body.message };
      }
      return { data: body };
    });
  },
  async update(cookies: any, user: any, ctx = null) {
    const token = tokenService.get(ctx);
    return HttpClient(`${process.env.NEXT_PUBLIC_URL_API}/users/update`, {
      method: "POST",
      headers: { "Authorization": `Bearer ${token}` },
      refresh: true,
      cookies,
      body: user,
    }).then(async (res) => {
      const body = await res.body;
      if (res.ok === false) {
        return { errors: body.message };
      }
      return { data: body };
    });
  },
};
