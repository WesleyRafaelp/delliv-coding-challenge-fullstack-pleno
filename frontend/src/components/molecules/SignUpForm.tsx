
import React, { useState } from 'react';
import Input from '../atoms/Input';
import Button from '../atoms/Button';
import { userService } from '@/services/user/user-service';
import { useForm } from 'react-hook-form';
import { CheckBadgeIcon } from '@heroicons/react/24/solid';

const SignUpForm: React.FC = () => {
  const {
    register,
    reset,
    handleSubmit,
    control,
    formState: { errors },
  } = useForm({
    criteriaMode: "all",
  });
  const [isFormSubmitted, setIsFormSubmitted] = useState(false);

  const onSubmit = async (data: any) => {
    const user = {
      name:data.name,
      email:data.email,
      password:data.password,
    }
    try {
      await userService.create(user);
      setIsFormSubmitted(true);
      reset();
      setTimeout(() => {
        setIsFormSubmitted(false);
      }, 5000); 
    } catch (err) {
      alert(`Usuário inválido: ${err}`);
    }
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} className="flex flex-col gap-4 bg-white border border-gray-300 p-4">
      <label className="block font-semibold text-gray-700">
        Nome de usuário:
        <Input type="text" placeholder='Digite seu nome' register={register('name', {required: 'Nome é obrigatório', minLength: {
          value: 8,
          message: 'A senha deve ter pelo menos 8 caracteres.'
        },
        maxLength: {
          value: 30,
          message: 'A senha não pode ter mais de 30 caracteres.'
        },})} />
      </label>
      {errors.name && (
            <p className="text-red-500">{`${errors.name.message}`}</p>
          )}
      <label className="block font-semibold text-gray-700">
        E-mail:
        <Input type="email" placeholder='Digite seu email' register={register('email', {required: 'E-mail é obrigatório'})} />
      </label>
      {errors.email && (
            <p className="text-red-500">{`${errors.email.message}`}</p>
          )}
      <label className="block font-semibold text-gray-700">
        Senha:
        <Input type="password" placeholder='Crie uma senha' register={register('password', {required: 'Senha é obrigatório', 
        minLength: {
          value: 8,
          message: 'A senha deve ter pelo menos 8 caracteres.'
        },
        maxLength: {
          value: 30,
          message: 'A senha não pode ter mais de 30 caracteres.'
        },
      validate: value => {
        const regex = /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])[a-zA-Z0-9!@#$%^&*]+$/;
        return regex.test(value) || 'A senha deve conter pelo menos uma letra maiúscula, uma letra minúscula e um número.';
      }})} />
      </label>
      {errors.password && (
            <p className="text-red-500">{`${errors.password.message}`}</p>
          )}
      <Button type='submit'>Cadastrar</Button>
      {isFormSubmitted ? (
        <div className='flex w-full justify-center items-center mt-2'>
          <p className="text-blue-700 font-bold text-center mb-4 flex m-auto">
      Usuário criado com sucesso !
          <CheckBadgeIcon className='w-4 text-blue-700'/>
          </p>
        </div>
      ) : null}
    </form>
  );
};

export default SignUpForm;
