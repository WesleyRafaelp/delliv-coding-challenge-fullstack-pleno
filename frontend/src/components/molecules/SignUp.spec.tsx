import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import SignUpForm from './SignUpForm';
import { userService } from '@/services/user/user-service';

describe('SignUpForm Component', () => {
  it('renders SignUpForm component', () => {
    render(<SignUpForm />);
    expect(screen.getByLabelText(/nome de usuário/i)).toBeInTheDocument();
  });

  it('submits the form and creates a new user', async () => {
    const mockCreate = jest.fn();
    userService.create = mockCreate;
    render(<SignUpForm />);

    fireEvent.change(screen.getByLabelText(/nome de usuário/i), { target: { value: 'JohnDoe123' } });
    fireEvent.change(screen.getByLabelText(/e-mail/i), { target: { value: 'john.doe@example.com' } });
    fireEvent.change(screen.getByLabelText(/senha/i), { target: { value: 'Password123' } });

    fireEvent.click(screen.getByText(/cadastrar/i));

    await waitFor(() => {
      expect(mockCreate).toHaveBeenCalledWith({
        name: 'JohnDoe123',
        email: 'john.doe@example.com',
        password: 'Password123',
      });
    });
  });

  it('displays error messages for invalid input', async () => {
    render(<SignUpForm />);

    fireEvent.click(screen.getByText(/cadastrar/i));

    await waitFor(() => {
      expect(screen.getByText(/Nome é obrigatório/i)).toBeInTheDocument();
      expect(screen.getByText(/E-mail é obrigatório/i)).toBeInTheDocument();
      expect(screen.getByText(/Senha é obrigatório/i)).toBeInTheDocument();
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<SignUpForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
