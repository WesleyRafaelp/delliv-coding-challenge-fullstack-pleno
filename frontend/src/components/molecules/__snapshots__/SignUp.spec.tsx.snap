// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm Component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col gap-4 bg-white border border-gray-300 p-4"
  >
    <label
      class="block font-semibold text-gray-700"
    >
      Nome de usu√°rio:
      <input
        class="border border-gray-300 rounded p-2 w-full focus:outline-none focus:ring focus:border-blue-300"
        name="name"
        placeholder="Digite seu nome"
        type="text"
        value=""
      />
    </label>
    <label
      class="block font-semibold text-gray-700"
    >
      E-mail:
      <input
        class="border border-gray-300 rounded p-2 w-full focus:outline-none focus:ring focus:border-blue-300"
        name="email"
        placeholder="Digite seu email"
        type="email"
        value=""
      />
    </label>
    <label
      class="block font-semibold text-gray-700"
    >
      Senha:
      <input
        class="border border-gray-300 rounded p-2 w-full focus:outline-none focus:ring focus:border-blue-300"
        name="password"
        placeholder="Crie uma senha"
        type="password"
        value=""
      />
    </label>
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline-blue active:bg-blue-800"
      type="submit"
    >
      Cadastrar
    </button>
  </form>
</DocumentFragment>
`;
